src/peer_connection.cpp

```diff
@@ -2993,6 +2993,15 @@ namespace libtorrent
 
 		TORRENT_ASSERT(m_outstanding_writing_bytes >= 0);
 
+		// every peer is entitled to allocate a disk buffer if it has no writes outstanding
+		// see the comment in incoming_piece
+		if (m_outstanding_writing_bytes == 0
+			&& m_channel_state[download_channel] & peer_info::bw_disk)
+		{
+			m_counters.inc_stats_counter(counters::num_peers_down_disk, -1);
+			m_channel_state[download_channel] &= ~peer_info::bw_disk;
```

Comment:
one thing I would be a little bit concerned about with this is that there may be other assumptions associated with the bw_disk bit being set. I believe it being set means that we expect to get a callback from the disk thread once there's enough space in the disk cache. And perhaps there's an assert in that callback to make sure this bit is still set. In the case where the cache size is 0 though, we probably won't ever get that callback, and this kind of fixes up that case.

It does open up for the edge case of changing the cache size, where we may hit this _and_ get the callback at some later point. (firing the assert)